### PR TITLE
Add navigation mesh source geometry parsers and callbacks

### PR DESCRIPTION
**Pull request by smix8**
Add navigation mesh source geometry parsers and callbacks so that externals can hook their own geometry into the internal navigation mesh baking process.

Proposal https://github.com/godotengine/godot-proposals/issues/5138

This can be used to hook addons, modules and extensions into the navigation mesh baking process.

E.g. Terrain3D or VoxelTool to name some examples. What all those extensions have currently in common is that they need to use suboptimal processes to add their geometry for navmesh baking and can't work frictionless with the default baking.

nav_baking

If a parser with callback is registered the `Callable` will be called for every single node that gets parsed when the NavigationServer `parse_source_geometry_data()` is used by something.

The Callable will get called with 3 args, the navigation mesh for the parse and bake properties, the source geometry object to add geometry to, and the actual node that gets parsed.

If the parser is no longer needed, e.g. because addon gets disabled, free the parser RID with the NavigationServer.

### 2D

```gdscript
var source_geometry_parser: RID
var source_geometry_parser_callback: Callable

func _ready() -> void:
	source_geometry_parser_callback = Callable(self, "_on_source_geometry_parser_callback")
	source_geometry_parser = NavigationServer2D.source_geometry_parser_create()
	NavigationServer2D.source_geometry_parser_set_callback(source_geometry_parser, source_geometry_parser_callback)

func _on_source_geometry_parser_callback(
	p_navigation_mesh: NavigationPolygon,
	p_source_geometry_data: NavigationMeshSourceGeometryData2D,
	p_parsed_node: Node) -> void:
	pass
	# Add custom source geometry to p_source_geometry_data.
	# The p_navigation_mesh holds the parse settings. Do NOT edit or add directly to p_navigation_mesh.
	# The p_parsed_node is the currently parsed node.

func delete_parser() -> void:
	NavigationServer2D.free_rid(source_geometry_parser)
	source_geometry_parser = RID()
```

### 3D

```gdscript
var source_geometry_parser: RID
var source_geometry_parser_callback: Callable

func _ready() -> void:
	source_geometry_parser_callback = Callable(self, "_on_source_geometry_parser_callback")
	source_geometry_parser = NavigationServer3D.source_geometry_parser_create()
	NavigationServer3D.source_geometry_parser_set_callback(source_geometry_parser, source_geometry_parser_callback)

func _on_source_geometry_parser_callback(
	p_navigation_mesh: NavigationMesh,
	p_source_geometry_data: NavigationMeshSourceGeometryData3D,
	p_parsed_node: Node) -> void:
	pass
	# Add custom source geometry to p_source_geometry_data.
	# The p_navigation_mesh holds the parse settings. Do NOT edit or add directly to p_navigation_mesh.
	# The p_parsed_node is the currently parsed node.

func delete_parser() -> void:
	NavigationServer3D.free_rid(source_geometry_parser)
	source_geometry_parser = RID()
```

### CODE REVIEW

**Comment by smix8** [comment]
@Zylann and @TokisanGames might be interested in (reviewing) this.

**Review by fire — APPROVED** [review]
I did not do a technical review.

> Would be good to squeeze into 4.3 as it would be many more months of hanging dry addons and extensions that can not really get their users to use the hot master build.

Would be good for gdextension authors as a quality of life enhancement.

**Comment by TokisanGames** [comment]
@tcoxon is our resident expert on navigation in Terrain3D.

File: modules/navigation/3d/nav_mesh_generator_3d.h [hunk 1]
```diff
@@ -47,6 +48,14 @@ class NavMeshGenerator3D : public Object {
 	static Mutex baking_navmesh_mutex;
 	static Mutex generator_task_mutex;
 
+	static RWLock generator_rid_rwlock;
+	struct NavMeshGeometryParser3D {
+		RID self;
+		Callable callback = Callable();
```
Comment (AThousandShips) [inline modules/navigation/3d/nav_mesh_generator_3d.h:54]:
```suggestion
		Callable callback;
```
Unnecessary default

File: modules/navigation/3d/nav_mesh_generator_3d.cpp [hunk 2]
```diff
@@ -920,4 +939,46 @@ bool NavMeshGenerator3D::generator_emit_callback(const Callable &p_callback) {
...
+void NavMeshGenerator3D::free(RID p_object) {
+	RWLockWrite write_lock(generator_rid_rwlock);
+
+	if (generator_parser_owner.owns(p_object)) {
+		NavMeshGeometryParser3D *parser = generator_parser_owner.get_or_null(p_object);
+
+		generator_parsers.erase(parser);
+
+		generator_parser_owner.free(p_object);
+
```
Comment (AThousandShips) [inline modules/navigation/3d/nav_mesh_generator_3d.cpp:978]:
```suggestion
```

File: modules/navigation/3d/godot_navigation_server_3d.cpp [hunk 3]
```diff
@@ -1428,6 +1433,23 @@ PathQueryResult GodotNavigationServer3D::_query_path(const PathQueryParameters &
...
+	return RID();
+};
+
+void GodotNavigationServer3D::source_geometry_parser_set_callback(RID p_parser, const Callable &p_callback) {
+#ifndef _3D_DISABLED
+	if (navmesh_generator_3d) {
+		navmesh_generator_3d->source_geometry_parser_set_callback(p_parser, p_callback);
+	}
+#endif // _3D_DISABLED
+};
```
Comment (AThousandShips) [inline modules/navigation/3d/godot_navigation_server_3d.cpp:1451]:
```suggestion
}
```

File: modules/navigation/3d/godot_navigation_server_3d.cpp [hunk 4]
```diff
@@ -1428,6 +1433,23 @@ PathQueryResult GodotNavigationServer3D::_query_path(const PathQueryParameters &
...
 }
 
+RID GodotNavigationServer3D::source_geometry_parser_create() {
+#ifndef _3D_DISABLED
+	if (navmesh_generator_3d) {
+		return navmesh_generator_3d->source_geometry_parser_create();
+	}
+#endif // _3D_DISABLED
+	return RID();
+};
```
Comment (AThousandShips) [inline modules/navigation/3d/godot_navigation_server_3d.cpp:1443]:
```suggestion
}
```

File: modules/navigation/2d/nav_mesh_generator_2d.h [hunk 5]
```diff
@@ -46,6 +47,14 @@ class NavMeshGenerator2D : public Object {
 	static Mutex baking_navmesh_mutex;
 	static Mutex generator_task_mutex;
 
+	static RWLock generator_rid_rwlock;
+	struct NavMeshGeometryParser2D {
+		RID self;
+		Callable callback = Callable();
```
Comment (AThousandShips) [inline modules/navigation/2d/nav_mesh_generator_2d.h:53]:
```suggestion
		Callable callback;
```

File: modules/navigation/2d/nav_mesh_generator_2d.cpp [hunk 6]
```diff
@@ -813,6 +832,48 @@ bool NavMeshGenerator2D::generator_emit_callback(const Callable &p_callback) {
...
+void NavMeshGenerator2D::free(RID p_object) {
+	RWLockWrite write_lock(generator_rid_rwlock);
+
+	if (generator_parser_owner.owns(p_object)) {
+		NavMeshGeometryParser2D *parser = generator_parser_owner.get_or_null(p_object);
+
+		generator_parsers.erase(parser);
+
+		generator_parser_owner.free(p_object);
+
```
Comment (AThousandShips) [inline modules/navigation/2d/nav_mesh_generator_2d.cpp:871]:
```suggestion
```

File: modules/navigation/2d/godot_navigation_server_2d.cpp [hunk 7]
```diff
@@ -453,3 +462,20 @@ void GodotNavigationServer2D::query_path(const Ref<NavigationPathQueryParameters
...
+	return RID();
+};
+
+void GodotNavigationServer2D::source_geometry_parser_set_callback(RID p_parser, const Callable &p_callback) {
+#ifdef CLIPPER2_ENABLED
+	if (navmesh_generator_2d) {
+		navmesh_generator_2d->source_geometry_parser_set_callback(p_parser, p_callback);
+	}
+#endif // CLIPPER2_ENABLED
+};
```
Comment (AThousandShips) [inline modules/navigation/2d/godot_navigation_server_2d.cpp:481]:
```suggestion
}
```

File: modules/navigation/2d/godot_navigation_server_2d.cpp [hunk 8]
```diff
@@ -453,3 +462,20 @@ void GodotNavigationServer2D::query_path(const Ref<NavigationPathQueryParameters
...
 }
+
+RID GodotNavigationServer2D::source_geometry_parser_create() {
+#ifdef CLIPPER2_ENABLED
+	if (navmesh_generator_2d) {
+		return navmesh_generator_2d->source_geometry_parser_create();
+	}
+#endif // CLIPPER2_ENABLED
+	return RID();
+};
```
Comment (AThousandShips) [inline modules/navigation/2d/godot_navigation_server_2d.cpp:473]:
```suggestion
}
```

File: modules/navigation/2d/nav_mesh_generator_2d.cpp [hunk 9]
```diff
@@ -236,6 +246,15 @@ void NavMeshGenerator2D::generator_parse_geometry_node(Ref<NavigationPolygon> p_
 	generator_parse_tilemap_node(p_navigation_mesh, p_source_geometry_data, p_node);
 	generator_parse_navigationobstacle_node(p_navigation_mesh, p_source_geometry_data, p_node);
 
+	generator_rid_rwlock.read_lock();
+	for (const NavMeshGeometryParser2D *parser : generator_parsers) {
+		if (parser->callback.is_null()) {
```
Comment (AThousandShips) [inline modules/navigation/2d/nav_mesh_generator_2d.cpp:251]:
```suggestion
		if (!parser->callback.is_valid()) {
```
`Callable` works differently I believe, needs to be `is_valid`

File: modules/navigation/3d/nav_mesh_generator_3d.cpp [hunk 10]
```diff
@@ -254,6 +264,15 @@ void NavMeshGenerator3D::generator_parse_geometry_node(const Ref<NavigationMesh>
 #endif
 	generator_parse_navigationobstacle_node(p_navigation_mesh, p_source_geometry_data, p_node);
 
+	generator_rid_rwlock.read_lock();
+	for (const NavMeshGeometryParser3D *parser : generator_parsers) {
+		if (parser->callback.is_null()) {
```
Comment (AThousandShips) [inline modules/navigation/3d/nav_mesh_generator_3d.cpp:269]:
```suggestion
		if (!parser->callback.is_valid()) {
```

**Review by AThousandShips — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**